datastore/mongodb/filters.go

```diff
@@ -141,3 +142,88 @@ func (c *Client) AddFilterDimension(ctx context.Context, fID string, dimension m
 	}
 	return nil
 }
+
+func (c *Client) UpdateFilterDimension(ctx context.Context, filterID string, dimensionName string, dimension model.Dimension, currentETag string) (string, error) {
+	col := c.collections.filters
+
+	lockID, err := col.lock(ctx, filterID)
+	if err != nil {
+		return "", errors.Wrap(err, "error while locking filters collection: %w")
+	}
+
+	defer col.unlock(ctx, lockID)
+
+	filter, err := c.GetFilter(ctx, filterID)
+	if err != nil {
+		return "", &er{
+			err:        errors.Wrapf(err, "unable to fetch filter (%s)", filterID),
+			badRequest: true,
+		}
+	}
+
+	// The Etag is optional, so we check if it's provided before comparing it.
+	if currentETag != "" && currentETag != filter.ETag {
+		return "", &er{
+			err:      errors.New("conflict: invalid ETag provided or filter has been updated"),
+			conflict: true,
+			logData: log.Data{
+				"expected_etag": currentETag,
+				"actual_etag":   filter.ETag,
+			},
+		}
+	}
+
+	// Find the dimension in order to replace it in-memory with the old one (rather than just
+	// in the datastore). We need to do this in order to generate an accurate ETag/hash.
+	newDimensionIndex, err := findDimensionIndex(filter, dimensionName)
+	if err != nil {
+		return "", &er{
+			err:      errors.Wrapf(err, "failed to find dimension (%s)", dimensionName),
+			notFound: true,
+		}
+	}
+
+	filter.Dimensions[newDimensionIndex] = dimension
+
+	if filter.ETag, err = filter.Hash(nil); err != nil {
+		return "", errors.Wrap(err, "failed to generate eTag")
+	}
+
+	selectFilter := bson.M{
+		"filter_id":       filterID,
+		"dimensions.name": dimensionName,
+	}
+
+	updateFilter := bson.M{
+		"$set": bson.M{
+			"etag":         filter.ETag,
+			"dimensions.$": dimension,
+		},
+	}
+
+	result, err := c.conn.Collection(col.name).Update(ctx, selectFilter, updateFilter)
+	if err != nil {
+		return "", errors.Wrapf(err, "failed to update filter (%s) dimension (%s)", filterID, dimensionName)
```

Comment:
Where you're using a lot of wrapf etc, would it be better to move some of that information into logData?